bug: grafana_role_assignment can not assign teams outside main org

### Terraform Version
* Terraform: 1.5.7
* Terraform Grafana Provider: 2.3.3
* Grafana: 10.1.2 (enterprise)

### Affected Resource(s)
grafana_role_assignment

### Terraform Configuration Files
```hcl
resource "grafana_team" "example" {
  org_id  = 2
  name    = "Example Team"
  members = []
}

resource "grafana_role_assignment" "example" {
  role_uid = "my_org_editor"
  teams    = [grafana_team.example.id]
}
```

### Terraform plan
```
Terraform will perform the following actions:

  # grafana_role_assignment.example will be created
  + resource "grafana_role_assignment" "example" {
      + id       = (known after apply)
      + role_uid = "my_org_editor"
      + teams    = (known after apply)
    }

  # grafana_team.example will be created
  + resource "grafana_team" "example" {
      + id                               = (known after apply)
      + ignore_externally_synced_members = true
      + name                             = "Example Team"
      + org_id                           = "2"
      + team_id                          = (known after apply)
    }

Plan: 2 to add, 0 to change, 0 to destroy.
```

### Terraform plan after retry
```
grafana_team.example: Refreshing state... [id=2:204]

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # grafana_role_assignment.example will be created
  + resource "grafana_role_assignment" "example" {
      + id       = (known after apply)
      + role_uid = "my_org_editor"
      + teams    = [
          + "2:204",
        ]
    }
```

### Debug Output
```
grafana_role_assignment.example: Still creating... [10s elapsed]
2023-10-12T20:45:03.505+0200 [ERROR] provider.terraform-provider-grafana_v2.3.3: Response contains error diagnostic: tf_proto_version=5.4 tf_provider_addr=registry.terraform.io/grafana/grafana tf_rpc=ApplyResourceChange @module=sdk.proto diagnostic_detail= diagnostic_summary="status: 500, body: {"message":"Failed to update role assignments","traceID":""}" tf_req_id=ca83d5bf-89bd-c7f0-f2c1-3cad8afb29d8 tf_resource_type=grafana_role_assignment @caller=github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/internal/diag/diagnostics.go:58 diagnostic_severity=ERROR timestamp=2023-10-12T20:45:03.505+0200
2023-10-12T20:45:03.614+0200 [DEBUG] State storage *statemgr.Filesystem declined to persist a state snapshot
2023-10-12T20:45:03.614+0200 [ERROR] vertex "grafana_role_assignment.example" error: status: 500, body: {"message":"Failed to update role assignments","traceID":""}

╷
│ Error: status: 500, body: {"message":"Failed to update role assignments","traceID":""}
│ 
│   with grafana_role_assignment.example,
│   on organizations.tf line 65, in resource "grafana_role_assignment" "example":
│   65: resource "grafana_role_assignment" "example" {
│ 
╵
2023-10-12T20:45:03.711+0200 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2023-10-12T20:45:03.712+0200 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/grafana/grafana/2.3.3/linux_amd64/terraform-provider-grafana_v2.3.3 pid=437573
2023-10-12T20:45:03.712+0200 [DEBUG] provider: plugin exited
```

### **Grafana log**
```
grafana-664f79fd67-zwjpg grafana logger=context userId=1 orgId=1 uname=admin t=2023-10-12T18:45:03.493576442Z level=error msg="Failed to update role assignments" error="could not verify the existence of the specified teams: resource does not exist" remote_addr=127.0.0.1 traceID=
grafana-664f79fd67-zwjpg grafana logger=context userId=1 orgId=1 uname=admin t=2023-10-12T18:45:03.493623094Z level=error msg="Request Completed" method=PUT path=/api/access-control/roles/my_org_editor/assignments status=500 remote_addr=127.0.0.1 time_ms=72 duration=72.646959ms size=60 referer= handler=/api/access-control/roles/:roleUID/assignments
```

### Expected Behavior
It should work.

### Actual Behavior
It doesn't work

### Important Factoids
It seems to me that the POST is done to the main org, where perhaps it should be done to org 2?

The role UID definitely exists because I can query it at /api/access-control/roles/my_org_editor . Also it is a global role.

--- a/docs/data-sources/role.md
+++ b/docs/data-sources/role.md
@@ -58,6 +58,7 @@ data "grafana_role" "from_name" {
 - `group` (String) Group of the role. Available with Grafana 8.5+.
 - `hidden` (Boolean) Boolean to state whether the role should be visible in the Grafana UI or not. Available with Grafana 8.5+.
 - `id` (String) The ID of this resource.
+- `org_id` (String) The Organization ID. If not set, the Org ID defined in the provider block will be used.
 - `permissions` (Set of Object) Specific set of actions granted by the role. (see [below for nested schema](#nestedatt--permissions))
 - `uid` (String) Unique identifier of the role. Used for assignments.
 - `version` (Number) Version of the role. A role is updated only on version increase.

--- a/docs/resources/role.md
+++ b/docs/resources/role.md
@@ -54,6 +54,7 @@ resource "grafana_role" "super_user" {
 - `global` (Boolean) Boolean to state whether the role is available across all organizations or not. Defaults to `false`.
 - `group` (String) Group of the role. Available with Grafana 8.5+.
 - `hidden` (Boolean) Boolean to state whether the role should be visible in the Grafana UI or not. Available with Grafana 8.5+. Defaults to `false`.
+- `org_id` (String) The Organization ID. If not set, the Org ID defined in the provider block will be used.
 - `permissions` (Block Set) Specific set of actions granted by the role. (see [below for nested schema](#nestedblock--permissions))
 - `uid` (String) Unique identifier of the role. Used for assignments.
 

--- a/docs/resources/role_assignment.md
+++ b/docs/resources/role_assignment.md
@@ -61,6 +61,7 @@ resource "grafana_role_assignment" "test" {
 
 ### Optional
 
+- `org_id` (String) The Organization ID. If not set, the Org ID defined in the provider block will be used.
 - `service_accounts` (Set of String) IDs of service accounts that the role should be assigned to.
 - `teams` (Set of String) IDs of teams that the role should be assigned to.
 - `users` (Set of Number) IDs of users that the role should be assigned to.

--- a/internal/resources/grafana/data_source_role.go
+++ b/internal/resources/grafana/data_source_role.go
@@ -28,7 +28,7 @@ func DatasourceRole() *schema.Resource {
 }
 
 func dataSourceRoleRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*common.Client).GrafanaAPI
+	client, orgID := ClientFromNewOrgResource(meta, d)
 	roles, err := client.GetRoles()
 	if err != nil {
 		return diag.FromErr(err)
@@ -37,6 +37,7 @@ func dataSourceRoleRead(ctx context.Context, d *schema.ResourceData, meta interf
 	name := d.Get("name").(string)
 	for _, r := range roles {
 		if r.Name == name {
+			d.SetId(MakeOrgResourceID(orgID, r.UID))
 			return readRoleFromUID(client, r.UID, d)
 		}
 	}

--- a/internal/resources/grafana/resource_role.go
+++ b/internal/resources/grafana/resource_role.go
@@ -26,6 +26,7 @@ func ResourceRole() *schema.Resource {
 			StateContext: schema.ImportStatePassthroughContext,
 		},
 		Schema: map[string]*schema.Schema{
+			"org_id": orgIDAttribute(),
 			"uid": {
 				Type:        schema.TypeString,
 				Computed:    true,
@@ -96,7 +97,7 @@ func ResourceRole() *schema.Resource {
 }
 
 func CreateRole(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*common.Client).GrafanaAPI
+	client, orgID := ClientFromNewOrgResource(meta, d)
 
 	role := gapi.Role{
 		UID:         d.Get("uid").(string),
@@ -113,12 +114,8 @@ func CreateRole(ctx context.Context, d *schema.ResourceData, meta interface{}) d
 	if err != nil {
 		return diag.FromErr(err)
 	}
-	err = d.Set("uid", r.UID)
-	if err != nil {
-		return diag.FromErr(err)
-	}
-	d.SetId(r.UID)
-	return nil
+	d.SetId(MakeOrgResourceID(orgID, r.UID))
+	return ReadRole(ctx, d, meta)
 }
 
 func permissions(d *schema.ResourceData) []gapi.Permission {
@@ -140,8 +137,8 @@ func permissions(d *schema.ResourceData) []gapi.Permission {
 }
 
 func ReadRole(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*common.Client).GrafanaAPI
-	return readRoleFromUID(client, d.Id(), d)
+	client, _, uid := ClientFromExistingOrgResource(meta, d.Id())
+	return readRoleFromUID(client, uid, d)
 }
 
 func readRoleFromUID(client *gapi.Client, uid string, d *schema.ResourceData) diag.Diagnostics {
@@ -194,18 +191,17 @@ func readRoleFromUID(client *gapi.Client, uid string, d *schema.ResourceData) di
 	if err != nil {
 		return diag.FromErr(err)
 	}
-	d.SetId(r.UID)
 
 	return nil
 }
 
 func UpdateRole(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*common.Client).GrafanaAPI
+	client, _, uid := ClientFromExistingOrgResource(meta, d.Id())
 
 	if d.HasChange("version") || d.HasChange("name") || d.HasChange("description") || d.HasChange("permissions") ||
 		d.HasChange("display_name") || d.HasChange("group") || d.HasChange("hidden") {
 		r := gapi.Role{
-			UID:         d.Id(),
+			UID:         uid,
 			Name:        d.Get("name").(string),
 			Global:      d.Get("global").(bool),
 			Description: d.Get("description").(string),
@@ -220,16 +216,11 @@ func UpdateRole(ctx context.Context, d *schema.ResourceData, meta interface{}) d
 		}
 	}
 
-	return nil
+	return ReadRole(ctx, d, meta)
 }
 
 func DeleteRole(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*common.Client).GrafanaAPI
-	uid := d.Id()
+	client, _, uid := ClientFromExistingOrgResource(meta, d.Id())
 	g := d.Get("global").(bool)
-
-	if err := client.DeleteRole(uid, g); err != nil {
-		return diag.FromErr(err)
-	}
-	return nil
+	return diag.FromErr(client.DeleteRole(uid, g))
 }

--- a/internal/resources/grafana/resource_role_assignment.go
+++ b/internal/resources/grafana/resource_role_assignment.go
@@ -26,6 +26,7 @@ func ResourceRoleAssignment() *schema.Resource {
 			StateContext: schema.ImportStatePassthroughContext,
 		},
 		Schema: map[string]*schema.Schema{
+			"org_id": orgIDAttribute(),
 			"role_uid": {
 				Type:        schema.TypeString,
 				Required:    true,
@@ -74,17 +75,13 @@ func ResourceRoleAssignment() *schema.Resource {
 }
 
 func ReadRoleAssignments(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*common.Client).GrafanaAPI
-	uid := d.Id()
+	client, _, uid := ClientFromExistingOrgResource(meta, d.Id())
 	assignments, err := client.GetRoleAssignments(uid)
 	if err, shouldReturn := common.CheckReadError("role assignments", d, err); shouldReturn {
 		return err
 	}
 
-	if err := setRoleAssignments(assignments, d); err != nil {
-		return diag.FromErr(err)
-	}
-	return nil
+	return diag.FromErr(setRoleAssignments(assignments, d))
 }
 
 func UpdateRoleAssignments(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
@@ -92,8 +89,7 @@ func UpdateRoleAssignments(ctx context.Context, d *schema.ResourceData, meta int
 		return nil
 	}
 
-	client := meta.(*common.Client).GrafanaAPI
-
+	client, orgID := ClientFromNewOrgResource(meta, d)
 	uid := d.Get("role_uid").(string)
 	users, err := collectRoleAssignmentsToFn(d.Get("users"))
 	if err != nil {
@@ -122,14 +118,13 @@ func UpdateRoleAssignments(ctx context.Context, d *schema.ResourceData, meta int
 		return diag.FromErr(err)
 	}
 
-	d.SetId(uid)
+	d.SetId(MakeOrgResourceID(orgID, uid))
 	return ReadRoleAssignments(ctx, d, meta)
 }
 
 func DeleteRoleAssignments(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*common.Client).GrafanaAPI
+	client, _, uid := ClientFromExistingOrgResource(meta, d.Id())
 
-	uid := d.Get("role_uid").(string)
 	ra := &gapi.RoleAssignments{
 		RoleUID:         uid,
 		Users:           []int{},
@@ -137,14 +132,11 @@ func DeleteRoleAssignments(ctx context.Context, d *schema.ResourceData, meta int
 		ServiceAccounts: []int{},
 	}
 
-	if _, err := client.UpdateRoleAssignments(ra); err != nil {
-		return diag.FromErr(err)
-	}
-	return nil
+	_, err := client.UpdateRoleAssignments(ra)
+	return diag.FromErr(err)
 }
 
 func setRoleAssignments(assignments *gapi.RoleAssignments, d *schema.ResourceData) error {
-	d.SetId(assignments.RoleUID)
 	if err := d.Set("role_uid", assignments.RoleUID); err != nil {
 		return err
 	}

--- a/internal/resources/grafana/resource_role_assignment_test.go
+++ b/internal/resources/grafana/resource_role_assignment_test.go
@@ -13,7 +13,7 @@ import (
 	"github.com/grafana/terraform-provider-grafana/internal/testutils"
 )
 
-func TestRoleAssignments(t *testing.T) {
+func TestAccRoleAssignments(t *testing.T) {
 	testutils.CheckEnterpriseTestsEnabled(t)
 	var roleAssignment gapi.RoleAssignments
 
@@ -39,6 +39,50 @@ func TestRoleAssignments(t *testing.T) {
 					resource.TestCheckResourceAttr(
 						"grafana_role_assignment.test", "teams.#", "1",
 					),
+				),
+			},
+			{
+				Config:  roleAssignmentConfig(testName),
+				Destroy: true,
+			},
+		},
+	})
+}
+
+func TestAccRoleAssignments_inOrg(t *testing.T) {
+	testutils.CheckEnterpriseTestsEnabled(t)
+	var roleAssignment gapi.RoleAssignments
+	var org gapi.Org
+
+	testName := acctest.RandString(10)
+
+	resource.ParallelTest(t, resource.TestCase{
+		ProviderFactories: testutils.ProviderFactories,
+		CheckDestroy:      testRoleAssignmentCheckDestroy(&roleAssignment),
+		Steps: []resource.TestStep{
+			{
+				Config: roleAssignmentConfigInOrg(testName),
+				Check: resource.ComposeTestCheckFunc(
+					testRoleAssignmentCheckExists("grafana_role_assignment.test", &roleAssignment),
+					resource.TestCheckResourceAttr(
+						"grafana_role_assignment.test", "role_uid", testName,
+					),
+					resource.TestCheckResourceAttr(
+						"grafana_role_assignment.test", "users.#", "2",
+					),
+					resource.TestCheckResourceAttr(
+						"grafana_role_assignment.test", "service_accounts.#", "1",
+					),
+					resource.TestCheckResourceAttr(
+						"grafana_role_assignment.test", "teams.#", "1",
+					),
+
+					// Check that the role is in the correct organization
+					resource.TestMatchResourceAttr("grafana_role.test", "id", nonDefaultOrgIDRegexp),
+					resource.TestMatchResourceAttr("grafana_role_assignment.test", "id", nonDefaultOrgIDRegexp),
+					testAccOrganizationCheckExists("grafana_organization.test", &org),
+					checkResourceIsInOrg("grafana_role.test", "grafana_organization.test"),
+					checkResourceIsInOrg("grafana_role_assignment.test", "grafana_organization.test"),
 				),
 			},
 			{
@@ -124,6 +168,61 @@ resource "grafana_role_assignment" "test" {
   users = [grafana_user.test_user.id, grafana_user.test_user2.id]
   teams = [grafana_team.test_team.id]
   service_accounts = [grafana_service_account.test.id]
+}
+`, name)
+}
+
+func roleAssignmentConfigInOrg(name string) string {
+	return fmt.Sprintf(`
+resource "grafana_organization" "test" {
+	name = "%[1]s"
+}
+
+resource "grafana_role" "test" {
+	org_id = grafana_organization.test.id
+
+	name  = "%[1]s"
+	description = "test desc"
+	version = 1
+	uid = "%[1]s"
+	global = true
+	group = "testgroup"
+	display_name = "testdisplay"
+	hidden = true
+  }
+
+resource "grafana_team" "test_team" {
+	org_id = grafana_organization.test.id
+	name = "%[1]s"
+}
+
+resource "grafana_user" "test_user" {
+	email = "%[1]s-1@test.com"
+	login    = "%[1]s-1@test.com"
+	password = "12345"
+}
+
+resource "grafana_user" "test_user2" {
+	email = "%[1]s-2@test.com"
+	login    = "%[1]s-2@test.com"
+	password = "12345"
+}
+
+resource "grafana_service_account" "test" {
+	org_id = grafana_organization.test.id
+
+	name        = "%[1]s-terraform-test"
+	role        = "Editor"
+	is_disabled = false
+  }
+
+resource "grafana_role_assignment" "test" {
+	org_id = grafana_organization.test.id
+
+  	role_uid = grafana_role.test.uid
+  	users = [grafana_user.test_user.id, grafana_user.test_user2.id]
+  	teams = [grafana_team.test_team.id]
+  	service_accounts = [grafana_service_account.test.id]
 }
 `, name)
 }

--- a/internal/resources/grafana/resource_role_test.go
+++ b/internal/resources/grafana/resource_role_test.go
@@ -9,6 +9,7 @@ import (
 
 	gapi "github.com/grafana/grafana-api-golang-client"
 	"github.com/grafana/terraform-provider-grafana/internal/common"
+	"github.com/grafana/terraform-provider-grafana/internal/resources/grafana"
 	"github.com/grafana/terraform-provider-grafana/internal/testutils"
 )
 
@@ -67,9 +68,19 @@ func testAccRoleCheckExists(rn string, r *gapi.Role) resource.TestCheckFunc {
 		if rs.Primary.ID == "" {
 			return fmt.Errorf("resource id not set")
 		}
-
+		orgID, roleUID := grafana.SplitOrgResourceID(rs.Primary.ID)
 		client := testutils.Provider.Meta().(*common.Client).GrafanaAPI
-		role, err := client.GetRole(rs.Primary.ID)
+
+		// If the org ID is set, check that the report doesn't exist in the default org
+		if orgID > 1 {
+			role, err := client.GetRole(roleUID)
+			if err == nil || role != nil {
+				return fmt.Errorf("expected no role with ID %s in default org but found one", roleUID)
+			}
+			client = client.WithOrgID(orgID)
+		}
+
+		role, err := client.GetRole(roleUID)
 		if err != nil {
 			return fmt.Errorf("error getting role: %s", err)
 		}

